pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java

```diff
@@ -154,12 +156,18 @@ public String getTypedId() {
     }
 
     /**
-     * Get attributes as immutable map.
+     * Get all attributes as immutable map.
      * 
      * @return the immutable attributes
      */
     public Map<String, Object> getAttributes() {
-        return Collections.unmodifiableMap(this.attributes);
+        final Map<String, Object> newAttributes = new HashMap<>();
+        for (Map.Entry<String, Object> entries : this.attributes.entrySet()) {
+            final String key = entries.getKey();
+            final Object value = getAttribute(key);
+            newAttributes.put(key, value);
+        }
+        return Collections.unmodifiableMap(newAttributes);
```

Comment:
I understand your concern and indeed, as I already create a new Map, this is not absolutely necessary.
An `unmodifiableMap` has always been returned by the `getAttributes` method and it has always worked with CAS. So I'd better keep it that way for backward compatibility, but this is something we can keep in mind and fix in a 2.0.1 at worst if necessary.